[c10d] Move sync_params to C++

goldsborough opened the pull request:
The next function I'm moving to C++ is `sync_params`. It is stacked on top of https://github.com/pytorch/pytorch/pull/9729, so some changes will go away when it lands and I rebase.

I also split code into a `.h` and `.cpp` file for better code organization.

@teng-li @pietern @apaszke

ezyang commented:
@pytorchbot retest this please

weiyangfb commented:
cc @teng-li would you like to review this PR too?

teng-li commented:
@goldsborough We do have a test covering this change right?

goldsborough commented:
@teng-li I ran the c10d DDP test that I wrote. Should we also write some unit tests for these individual functions?

teng-li commented on test/test_c10d.py:
```diff
@@ -509,10 +515,103 @@ def test_gloo_backend(self):
+        process_group = c10d.ProcessGroupGloo(store, self.rank, self.world_size, options)
+
+        # Use all available devices on every process here (data is small, so should be fine).
+        devices = list(range(torch.cuda.device_count()))
```
I am thinking of dividing the devices among all processes, wdyt?

goldsborough replied:
We do this in other tests, but in a lot of cases we only run with two GPUs, so with two processes that would only leave one device per process and make this test less meaningful. Is this reasonable?

teng-li replied:
it should be fine for Gloo, but NCCL it won't work AFAIK. We can leave as is

teng-li commented on test/test_c10d.py:
```diff
@@ -509,10 +515,103 @@ def test_gloo_backend(self):
+            parameter_data=parameter_data,
+            buffer_data=buffer_data,
+            devices=devices,
+            broadcast_bucket_size=25 * MB,
```
will 25MB be too big in this case? Ideally, we would like to have more than 1 bucket to be created.

goldsborough replied:
Good point

teng-li reviewed (approved):
Look great, thank you for adding those test cases.

Please see my comments.

goldsborough commented:
@apaszke @ezyang @soumith looking for codeowner approval

goldsborough commented:
@pytorchbot retest this please

goldsborough committed:
Write sync_params in C++ and move code into .cpp file

Rebase fixes

Rebase

Wrote unit test for dist_broadcast_coalesced

Wrote unit test for sync_params

Make the bucke size 20 bytes and increase test timeout to 10s

5 tensors per device

Increase timeout

goldsborough commented:
@pytorchbot retest this please

soumith commented:
might possibly need a rebase. cc: @teng-li

yf225 commented:
@pytorchbot retest this please

goldsborough commented:
@pytorchbot retest this please

goldsborough committed:
Debugging

goldsborough committed:
Use fewer devices

goldsborough committed:
More debuginfo

goldsborough committed:
Remove debug info
